File: mac/Sources/CodeBurnMenubar/SessionCountLabel.swift
/// Shared session-count phrasing. Keep in lockstep with `src/session-count-label.ts`.
enum SessionCountLabel {
    static var help: String { L("Older session logs may be unavailable.") }
    /// Combined-scope counts are a per-device numeric sum with no shared identity.
    /// Do not show that sum as unique or as a lower bound.
    static var combinedHelp: String { L("Session identities are unavailable across devices.") }
    static var combinedText: String { L("Session count unavailable") }

    static func isExact(_ basis: String?) -> Bool {
        basis == "identity"
    }

    static func text(sessions: Int, basis: String?) -> String {
        if !isExact(basis) {
            if sessions <= 0 { return combinedText }
            return sessions == 1 ? L("At least 1 session") : L("At least %lld sessions", sessions)
        }
        return sessions == 1 ? L("1 session") : L("%lld sessions", sessions)
    }

    static func compact(sessions: Int, basis: String?) -> String {
        if !isExact(basis) {
            if sessions <= 0 { return L("Unavailable") }
            return L("≥%lld sess", sessions)
        }
        return L("%lld sess", sessions)
    }

    static func averageText(_ value: Double?, basis: String?, format: (Double) -> String) -> String {
        guard isExact(basis), let value, value.isFinite else { return "—" }
        return format(value)
    }

    static func helpText(combined: Bool, basis: String?) -> String {
        if combined { return combinedHelp }
        return isExact(basis) ? "" : help
    }
}

File: app/renderer/lib/session-count-label.ts
/// Display copy for period session counts. Keep in lockstep with
/// `src/session-count-label.ts` (renderer does not import `src/`).
import { currentLocaleTag } from './i18n/index'

export type SessionCountBasis = 'identity' | 'partial'

export const SESSION_COUNT_HELP = 'Older session logs may be unavailable.'
export const COMBINED_SESSION_COUNT_HELP = 'Session identities are unavailable across devices.'
export const COMBINED_SESSION_COUNT_LABEL = 'Session count unavailable'

export function sessionCountIsExact(basis: SessionCountBasis | undefined): boolean {
  return basis === 'identity'
}

export function formatSessionCount(
  sessions: number,
  basis: SessionCountBasis | undefined,
): string {
  if (!sessionCountIsExact(basis)) {
    if (sessions <= 0) return 'Session count unavailable'
    return sessions === 1 ? 'At least 1 session' : `At least ${sessions.toLocaleString(currentLocaleTag())} sessions`
  }
  if (sessions === 1) return '1 session'
  return `${sessions.toLocaleString(currentLocaleTag())} sessions`
}

export function formatCombinedSessionCount(): string {
  return COMBINED_SESSION_COUNT_LABEL
}
